src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -21,18 +21,28 @@ public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> fir
                 throw Error.ArgumentNull(nameof(second));
             }
 
+            var firstCollection = first as ICollection<TSource>;
+            if (firstCollection != null)
+            {
+                var secondCollection = second as ICollection<TSource>;
+                if (secondCollection != null)
+                {
+                    return new Concat2CollectionIterator<TSource>(firstCollection, secondCollection);
+                }
+            }
+
             var concatFirst = first as ConcatIterator<TSource>;
```

Comment:
If ```first``` is an ```ICollection<TSource>```, it's not going to be a ```ConcatIterator<TSource>```, right?  Seems then like this can be restructured as:
```C#
Debug.Assert(!((first is ICollection<TSource>) && (first is ConcatIterator<TSource>)),
    "Didn't expect enumerable to be both a collection and a concat iterator");

var firstCollection = first as ICollection<TSource>;
if (firstCollection != null)
{
    var secondCollection = second as ICollection<TSource>;
    if (secondCollection != null)
    {
        return new Concat2CollectionIterator<TSource>(firstCollection, secondCollection);
    }
}
else
{
    var firstConcat = first as ConcatIterator<TSource>;
    if (firstConcat != null)
    {
        return firstConcat.Concat(second);
    }
}

return new Concat2EnumerableIterator<TSource>(first, second);
````
This avoids a type check and a branch when concatenating a collection with a non-collection. (Also note I renamed concatFirst to be firstConcat to match the earlier naming; otherwise it's strange to see some things named firstXx and other things xxFirst.)